File: python/src/ouster/sdk/mapping/slam_engine.py
from typing import List, Optional
from ouster.sdk.core import SensorInfo, LidarScan
from ouster.sdk.mapping.slam_backend import SlamConfig
from ouster.sdk.mapping.kiss_slam import KissSlam


class SlamEngine():
    """base class of general slam solutions for SDK cli usage"""

    def __init__(self, infos: List[SensorInfo], config: SlamConfig):
        """SlamEngine constructor"""
        if config.backend not in ["kiss"]:
            raise ValueError(f"Unsupported backend {config.backend}")
        self._backend = KissSlam(infos, config)

    def update(self, scans: List[Optional[LidarScan]]) -> List[Optional[LidarScan]]:
        """
        Update the pose (per_column_global_pose) variable in scan and return
        """
        return self._backend.update(scans)

    @property
    def last_pose(self):
        """Get the current pose"""
        return self._backend._last_slam_pose

File: ouster_mapping/src/slam_engine.cpp
#include <ouster/mapping/slam_engine.h>

#include <stdexcept>

#include "lio_slam.h"

namespace ouster {
namespace sdk {
namespace mapping {

std::unique_ptr<SlamEngine> SlamEngine::create(
    const std::vector<std::shared_ptr<core::SensorInfo>>& infos, const LIOSlamConfig& config) {
    return std::make_unique<LIOSlam>(infos, config);
}

}  // namespace mapping
}  // namespace sdk
}  // namespace ouster
